### ISSUE TITLE
T21: Run BFWS process-SFT and references and adjudicate the structural gate

### ISSUE THREAD

**Issue by Sino-Huang** [Sino-Huang/multimodality_on_planning#59]
> *This was generated by AI during triage.*

**Parent:** #38
**Plan ticket:** T21

## What to build
Run the pretrained base, process-SFT, random-valid, and exact-classical IW conditions, then adjudicate the structural gate and issue its receipt.

## Acceptance criteria
- All four IW conditions complete with full episode evidence.
- The structural gate issues an explicit adjudication receipt.
- No run starts without the matching authorization and gate receipt; outcomes distinguish PASS, VALID_STOP, INVALID, and ANCESTOR_STOP, with VALID_STOP and ANCESTOR_STOP producing a gated-not-run receipt and INVALID never treated as scientific completion.

## Blocked by
#52, #58

---
**Governed vocabulary reference:** This ticket uses the governed vocabulary (operational competence, structural/process competence, validity, stop outcome, text-state, visual-state, multimodal-state) exactly as defined in `CONTEXT.md` section "Governed downstream vocabulary" (lines 45-77), governed by issue #39 (T01). Terms are reused verbatim; no local redefinition applies.

**Comment by Sino-Huang** [comment]
Apply the resource-bounded evaluation protocol established by #54:

- drive incremental episode sessions in deterministic rounds and use batched Transformers generation with one shared backbone per GPU, adapter switching, adapter-isolated caching, and at most one request per active episode per round;
- freeze a 2x matching-reference decision-call limit separately from expansion limits; exhaustion is a valid unsuccessful episode;
- evaluate only final checkpoints across five seeds, with base, learned, random-valid, and exact-reference conditions on the identical selected tasks;
- qualify throughput without success outcomes before rollout, using the actual precision/batch settings and a cost-qualified fallback panel if full coverage cannot fit;
- stop before new launches at the wall-clock cutoff, preserve completed episodes, replay every completed episode, and never let partial coverage satisfy the gate.

Report invariant-valid success, invalid-operation rate, budget usage, and learned-vs-best-control gain separately. #54 reached 100% learned success on its reduced panel but still correctly `VALID_STOP` because random-valid also reached 100%; avoid interpreting panel success alone as structural advantage.

**Comment by Sino-Huang** [comment]
Dependency update from completed #56 (`62e02bc`): the governed arm is BFWS, not IW. Development references and the structural gate require the #56 authorization receipt, the 35 frozen dev tasks, final checkpoints only across seeds 17/29/43/71/101, and a per-episode model-call limit of twice the matching exact-reference decision count.

Before any model rollout, run outcome-blind hardware qualification using the frozen float32 deterministic batched schedule, adapter-isolated cache, scalar/batch plus repeated-batch probes, full coverage first and then only the preregistered exact-cost panel. Start one 20-hour gate clock after qualification, stop new calls at 18 hours, and require complete selected coverage. The frozen 45-task fresh test manifest is not authorized for this development gate; a successor authorization is required before efficacy access.

**Comment by Sino-Huang** [comment]
Dependency #58 is complete on main in commits `ad35bbe` and `cdd0bed`.

Corpus-ready retained inputs:
- `data/bfws_phase_v1/corpus-release/training/manifest.json`: ms-swift `messages` JSONL shards with 47,780 train and 21,239 dev process examples;
- `data/bfws_phase_v1/corpus-release/manifests/bfws-text-corpus.json`: 69,019 atomic process records, 67,215 operational records, 105 immutable semantic-task split assignments, and both views across all 35 strata;
- `data/bfws_phase_v1/corpus-release/audits/corpus.json`: zero semantic/input/input-target overlap, zero conflicting identical inputs, zero future leakage, zero live/corpus mismatch, zero parse/runtime/token rejection, and zero held-out access;
- `data/bfws_phase_v1/execution-receipts/generation-run-issue-56-bfws-development-v1-issue-58-bfws-text-corpus-v1-resume-004.json`: governed scientific-completion PASS with `byte_identical_regeneration: true`.

The process projection uses the shared `bounded_bfws_search_memory_v1` builder and BFWS chat serializer, 16 accepted deltas, 7,808 input tokens, 384 output tokens, and the pinned Qwen revision. Observed maxima are 7,360 input and 96 target tokens. The fresh 45-task test manifest was not opened or included. No learning command was run.

Operator details and resumable progress/ETA commands are in `docs/issue-58-bfws-text-corpus.md`. Full repository verification after review fixes: 662 passed; targeted Ruff and basedpyright clean. Standards and Spec reviews found no remaining actionable defects.

**Comment by Sino-Huang** [comment]
Implementation is pushed to `main` in `331a416` (`feat(planning): add governed BFWS structural gate runner`).

Implemented:
- one governed CLI for preflight, exact/random-valid references, five-seed process-only ms-swift LoRA SFT, outcome-blind two-GPU qualification, deterministic batched base/process rollout, independent replay, and structural-gate adjudication;
- exact issue-57 evidence plus issue-58 atomic process shards only: 47,780 train, 21,239 dev, 35 frozen dev tasks, and 21,239 matching exact decisions;
- shared `bounded_bfws_search_memory_v1` live input contract with 16 deltas and 7,808/384 token limits;
- matching exact expansion limits plus independent 2x exact-decision model-call limits;
- float32 batching with one backbone per GPU, isolated adapter caches, one request per active episode per round, scalar/batch and repeated-batch probes;
- full 35-task coverage first, then the preregistered cheapest-exact-cost task per domain (15 tasks, 27,228 unique scheduled calls) without using model outcomes;
- 18-hour launch cutoff, 20-hour gate deadline, resumable atomic gzip episode evidence, progress/elapsed/ETA logs, and independent reconstruction of all 16 selected episode products per task before adjudication;
- governed PASS, VALID_STOP, INVALID, and ANCESTOR_STOP receipts. The fresh 45-task held-out manifest is never opened.

Agent verification only; no learning or model rollout was run:
- `python scripts/run_bfws_issue59.py all --dry-run`: PASS, resolving 5 training launches and 2 GPU rollout shards;
- generated seed-17 smoke vector resolved successfully through the installed ms-swift 4.2.2 launcher help path;
- pinned Qwen processor accepted a real two-snapshot batch at 1,756 tokens;
- focused BFWS checks: 19 passed;
- full repository: 670 passed, one existing Albumentations update warning;
- targeted Ruff and basedpyright: clean.

Human command:
```bash
source ~/cd_vlaplan
python scripts/run_bfws_issue59.py all --devices cuda:0 cuda:1
```

Every long stage prints terminal progress with elapsed time and ETA. Stage-by-stage and resume commands are documented in `docs/issue-59-bfws-structural-gate.md`.

Issue #59 remains open because the actual five-seed training, qualification, rollout, replay, and adjudication products do not exist yet; the implementation does not claim an empirical result.

**Comment by Sino-Huang** [comment]
Performance/resume fix pushed in `ac71aa9` (`perf(planning): parallelize resumable BFWS references`).

Diagnosis confirmed the reference stage was serial: one main thread consumed ~99% of one CPU core while all other threads were idle. The fix now:
- runs exact replay and random-valid task/seed jobs in a bounded process pool;
- defaults to 8 workers and rejects a count above the process CPU affinity;
- preserves deterministic frozen manifest order regardless of completion order;
- keeps atomic per-episode gzip evidence;
- reopens and independently replays an existing episode, but never regenerates or overwrites it;
- generates only missing task/seed evidence on `--resume`;
- labels terminal progress as `generated`, `reused`, or `replayed`.

The interrupted run retained 149/175 random-valid episode files and no final manifest. A real four-job process-pool check passed: first pass statuses were exact `replayed` plus three `generated`; second pass was exact `replayed` plus three `reused`, with unchanged retained bytes and deterministic record order.

Resume command:
```bash
source ~/cd_vlaplan
python scripts/run_bfws_issue59.py all \
  --devices cuda:0 cuda:1 \
  --reference-workers 8 \
  --resume
```

Verification: 21 focused BFWS checks passed; full repository 672 passed with the one existing Albumentations warning; targeted Ruff and basedpyright clean. No partial issue-59 evidence was removed or rewritten by the implementation work.

**Comment by Sino-Huang** [comment]
Concurrent-GPU fix pushed in `59b1db7` (`fix(planning): isolate concurrent training ports`).

Root cause: seeds 17 and 29 were assigned to GPUs 0 and 1 but both torchrun launchers used port 29500. Seed 29 failed with `EADDRINUSE`, so the GPU-1 queue stopped.

Each frozen seed now retains a complete launch environment with distinct ports: seeds 17/29/43/71/101 use `MASTER_PORT` 29600/29601/29602/29603/29604 while GPU assignments remain 0/1/0/1/0. The actual subprocess environment is populated from that exact dry-run plan.

The real dry-run against the interrupted output passed, selecting attempt-002 for seeds 17 and 29. Verification: 22 focused BFWS checks, full repository 673 passed, and targeted Ruff/basedpyright clean.

Resume:
```bash
source ~/cd_vlaplan
python scripts/run_bfws_issue59.py all --devices cuda:0 cuda:1 --reference-workers 8 --resume
```

Local `AGENTS.md` now records the general A* safeguard: concurrent torchrun/ms-swift jobs need distinct explicit `MASTER_PORT` values; `CUDA_VISIBLE_DEVICES` does not isolate rendezvous ports.

**Comment by Sino-Huang** [comment]
Supervisor budget correction pushed in `8eb7abe` (`fix(planning): honor single-training BFWS budget`).

The earlier implementation interpreted the issue-56 five-seed freeze as five independent LoRA training runs. The supervisor explicitly overrides that for issue 59: train exactly one model because of time/compute limits.

New governed contract `issue-59-bfws-single-training-v1`:
- one process-SFT training run only;
- training seed 17 on physical `cuda:1`, world size 1;
- new immutable attempt is seed-17 attempt-003;
- one learned checkpoint and one matched base rollout seed;
- random-valid seed 17 is used for the matched adjudication (the already-retained other random-valid references remain reusable evidence but are not part of this comparison);
- qualification schedules two unique model sessions per task: base plus the single learned adapter;
- uncertainty remains the whole-problem-instance bootstrap; training-seed variance is explicitly not estimated;
- both GPUs remain available for the later two rollout shards.

The two stopped training attempts reached steps 1400 and 1388 of 4482 but produced no epoch checkpoint, so neither can be resumed as a model checkpoint. Both remain retained failed attempts.

Real dry-run passed: exactly one launch, seed 17, `CUDA_VISIBLE_DEVICES=1`, `MASTER_PORT=29600`, attempt-003, and qualification points to attempt-003 checkpoint-4482. Verification: 23 focused BFWS checks; full repository 674 passed; Ruff and basedpyright clean.

Run:
```bash
source ~/cd_vlaplan
python scripts/run_bfws_issue59.py all --devices cuda:0 cuda:1 --reference-workers 8 --resume
```

**Comment by Sino-Huang** [comment]
Implemented and pushed true BFWS training checkpoint resume in cee16f5.

Changes:
- `--resume` now discovers the newest complete seed-17 ms-swift checkpoint across preserved attempts and passes it via `--resume_from_checkpoint` to a new numbered attempt.
- Incomplete/partially written checkpoints and smoke checkpoints are ignored.
- Checkpoints are saved every half epoch: 747 optimizer steps.
- Supervisor budget override is now `issue-59-bfws-single-training-v2`: one seed-17 run, physical `cuda:1`, 2 epochs / 2,988 total optimizer steps.
- Qualification now expects `checkpoint-2988` rather than the old hard-coded `checkpoint-4482`.
- The runner continues to tee training output to the terminal and retained log.

Verification: 13 focused issue-59 tests passed; full suite 675 passed; Ruff and basedpyright clean. The real output tree currently has no valid checkpoint in attempts 001-003, so attempt 004 must start at step 0. Once it reaches step 747, later invocations with `--resume` can recover from that checkpoint.

Operator command:
```bash
source ~/cd_vlaplan
python scripts/run_bfws_issue59.py all \
  --devices cuda:0 cuda:1 \
  --reference-workers 8 \
  --resume
```
Training uses only `cuda:1`; both devices remain listed because later qualification/evaluation use both GPUs. Issue remains open pending the actual experiment and adjudication.

**Comment by Sino-Huang** [comment]
The authorized issue-59 workflow has completed and its terminal evidence is retained in `d1a6a3c` (`docs(planning): record BFWS gate valid stop`).

Final verified status:
- process exited; no issue-59 training/inference process remains;
- references complete: 35 exact BFWS plus 175 random-valid records (210 total);
- seed-17 two-epoch process-SFT training completed with `PASS` at 2,988 optimizer steps;
- complete checkpoints exist at steps 747, 1,494, 2,241, and 2,988;
- final learned checkpoint: `outputs/bfws_phase/issue59-v1/training/seed-17-attempt-004/checkpoints/checkpoint-2988`;
- outcome-blind qualification measured lower-95% throughput of 0.180490565 calls/s;
- the preregistered 9,076-call exact-cost panel projected 60,424.319 seconds after the frozen safety margin, exceeding the 15-hour rollout-certification budget;
- qualification therefore emitted `VALID_STOP`, selected zero tasks, and observed no model outcomes;
- no base/learned/random-valid/exact rollout matrix ran, so this attempt contains no BFWS SFT performance result;
- adjudication emitted the correct `VALID_STOP` gated-not-run receipt with `scientific_completion: false` and `start_permitted: false`.

Tracked terminal records:
- `data/bfws_phase_v1/issue59-terminal/training-report.json`
- `data/bfws_phase_v1/issue59-terminal/qualification.json`
- `data/bfws_phase_v1/issue59-terminal/adjudication-report.json`
- `data/bfws_phase_v1/issue59-terminal/gate-receipt.json`
- bounded interpretation in `docs/issue-59-bfws-structural-gate.md`

Verification: all four tracked records are byte-identical to the execution outputs; focused issue-59 suite: 13 passed. The 2 GB optimizer-bearing checkpoint remains in the ignored execution tree and was intentionally not committed.

This closure records a valid terminal resource stop, not structural-gate PASS or learned-policy failure. Any downstream scientific execution requiring a PASS ancestor must remain gated.

**Comment by Sino-Huang** [comment]
Distributed four-GPU replacement evaluation completed with terminal `PASS`; exact evidence is retained in `e61d7ee` (`docs(planning): record distributed BFWS gate pass`).

Final bounded result:
- merged outcome-blind qualification: `PASS` on the preregistered 15-domain exact-cost panel;
- four global rollout shards: all `PASS`, 30/30 model episodes complete;
- independent adjudication replay: 60/60 exact, random-valid, base, and process-SFT episodes;
- gate: `PASS`, `scientific_completion: true`;
- exact BFWS invariant-valid success: 1.0;
- process-SFT invariant-valid success: 0.933333;
- random-valid invariant-valid success: 0.466667;
- pretrained-base invariant-valid success: 0;
- absolute process-SFT gain over best control: 0.466667;
- paired whole-instance bootstrap gain lower bound: 0.2;
- process-SFT invalid-operation rate: 0.000452694;
- all four frozen threshold checks passed.

The result is limited to the 15 preregistered development tasks and one two-epoch seed-17 checkpoint. It does not estimate training-seed variance and did not access the fresh held-out test set.

Tracked records are under `data/bfws_phase_v1/issue59-distributed-terminal/`; the bounded interpretation is in `docs/issue-59-bfws-structural-gate.md`. The ticket remains closed with its final PASS result recorded.
